llama.cpp

```diff
@@ -15540,6 +15540,26 @@ float * llama_get_logits_ith(struct llama_context * ctx, int32_t i) {
     }
 }
 
+int32_t llama_get_logits_last_idx(struct llama_context * ctx) {
+    llama_synchronize(ctx);
+
+    if ((ctx->cparams.n_batch == 1) && (ctx->n_outputs == 1)) {
+        // trivial case, one input/output, return it.
+        return 0;
+    }
+
+    // non-trivial case scan for the last output.
+    for (int32_t i = ctx->output_ids.size() - 1; i >= 0; --i) {
+        const int32_t candidate = ctx->output_ids[i];
+        if (candidate >= 0) {
+            return i;
+        }
+    }
```

Comment:
I'm pretty sure there's a faster and simpler way to do this. If `ctx->n_outputs` is set to the total number of outputs after a batch is processed, the last logits would be available at `ctx->logits[ctx->n_outputs - 1]`.

(this could be a special case of `llama_get_logits_ith` instead, to avoid trying to find the last logits index, which is still non-trivial)

In `llama_decode_internal`, the number of outputs is already calculated as `n_outputs`, so a simple `lctx.n_outputs = n_outputs` near the end of the function (way after the compute graph and its inputs are built, outside of the `ubatch` loop) would do the trick.

(BTW, I worked on adding `ctx->output_ids`, `ctx->n_outputs`, and other related stuff in #6122)

Reply:
That seems reasonable to me!

I had been worried about changing the public api of `llama_get_logits_ith`, but if that is viable, then I concur, it seems guaranteed optimal to do it there, rather than on the index.

I had been hopeful that the worst case was uncommon, as it seemed likely that the last index would be set for wanting logits, and therefore the for loop would, in practice, usually exit on the first loop.

Would you like me to update this PR to conform to the suggested code?  I think the suggestion looks great, and even adds a cool piece of functionality of python-style negative indexing.

Reply:
> I had been worried about changing the public api of `llama_get_logits_ith`

Since this seems like a backward-compatible change, there is no need to worry. API changes are quite common in `llama.cpp`, there's even a "Recent API changes" section in the `README.md` for this purpose.

> I had been hopeful that the worst case was uncommon, as it seemed likely that the last index would be set for wanting logits, and therefore the for loop would, in practice, usually exit on the first loop.

You are right, the worst case isn't common (at least when using `batch.logits`). Another way would have been to always return `ctx->cparams.n_batch - 1` or `ctx->output_ids.size() - 1`, and let `llama_get_logits_ith` do its asserts to fail in unexpected cases, but this would not work with `llama_batch_get_one`.

Thinking a bit more about this, the worst case is actually quite common in the sense that it happens all the time when using `llama_batch_get_one`, because the "last valid index" in this case is 0.

> Would you like me to update this PR to conform to the suggested code?

Yes. You had the idea of making this, so feel free to make commits in your name.

I hereby allow @TheFlipbook to use my code review suggestion from this comment thread and do absolutely anything with it, including but not limited to committing it (even with modifications) without attributing me.

Reply:
Updated the PR with these suggestions.   Apologies for the double-push, the first one had a formatting mistake.

Let me know if you'd suggest any further updates!  Thanks for all the help!